rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,956 @@
+Scoped admins can create new roles, and access lists, however only if those roles have `grantable_scopes` and access lists have `scope` equal or more specific 
+than the scope of the granted roles of admins.
+
+Let’s say Alice would like to delegate admin rights for the lab environment to Ketanji and Bob:
+
+```yaml
+kind: access_list
+metadata:
+  name: scoped-dev-admin-dev
+spec:
+  scope: '/dev/lab`
+  grants: 
+  roles: [editor, auditor]
+  members:
+    - bob@example.com
+    - ketanji@example.com
+```
+
+Ketanji can now create roles, access lists and join tokens, but they all will have scopes set to /dev/lab
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: impersonator
+spec:
+  grantable_scope: '/dev/lab'
+  parent_resource_group: '/dev/lab' 
+  allow:
+    impersonate:
+      users: ['jenkins']
+      roles: ['jenkins']
+---
+kind: access_list
+metadata:
+  name: scoped-dev-admin-dev-access
+spec:
+  parent_resource_group: '/dev/lab'
+  scope: '/dev/lab`
+  grants: 
+  roles: [access]
+  members:
+    - bob@example.com
+    - ketanji@example.com
+---
+# token.yaml
+kind: token
+version: v2
+metadata:
+  name: my-token-name
+spec:
+  parent_resource_group: '/dev/lab'
+  roles: 
+    - Node
+    - App
+```
+
+By propagating scope for any resource created by a user leveraging the roles the granted scope, 
+we  make sure that our security invariant remains - Ketanji and Bob can’t expand the scope of their cluster access beyond /dev/lab.
+
+### Allow agent forwarding for some hosts
+
+Issue https://github.com/gravitational/teleport/issues/23790 is looking for a way to allow agent forwarding for specific hosts.
+
+This can be done with access lists and scoped assignment:
+
+```yaml
+kind: access_list
+metadata:
+  name: access-to-default
+spec:
+  grants: 
+    roles: [access-with-agent-forward]
+  scope:  '/dev/lab`
+  members:
+    - bob@example.com
+```
+
+This will let bob to ssh with agent forwarding on into hosts in scope of /dev/lab only.
+
+## Implementation
+
+Teleport will stop issuing new certificate extensions for access granted by Access Lists or Access Requests. Instead, Teleport will create and distribute internal `Grant` resources. These resources will be used by Teleport only and will never be exposed to users.
+
+Here is a high-level structure of an internal `grant` resource:
+
+```yaml
+kind: grant
+version: v1
+metadata:
+  # name is a unique random identifier uuid-v4 of this grant 
+  name: uuid
+spec:
+  # the time when the grant expires in RFC3339 format
+  create_time: "2023-02-22T21:04:26.312862Z"
+  # the time when the grant has been updated in RFC3339 format
+  update_time: "2023-02-22T21:04:26.312862Z"
+  # the time when the grant expires in RFC3339 format
+  expires: "2021-08-14T22:27:00Z"
+  # roles that are granted
+  roles: [admin, editor]
+  # traits that are granted
+  traits:
+    - 'internal.logins': 'root'
+  identity: 'alice@example.com'
```

Comment:
I worry about the implications of attaching this to the user globally. It's a big change from the certificate-based access we do today. If I am logged in to web UI and `tsh` on two different machines, I suppose access requests on one will change the access of the other.

This will hurt with resource access requests which currently are not additive - you can only have one at a time and you can only access the specifically requested resources while the request is "assumed". It will be necessary to solve this first (I believe we can).

Reply:
good point, I addressed your comment in `grants and access requests` section.